3D-microservices-observability/go-chi/services/shopping/shopping.go

```diff
@@ -0,0 +1,80 @@
+// 2019. Mahesh Voleti (mvoleti@vmware.com)
+
+package shopping
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+
+	"github.com/go-chi/chi"
+
+	. "wavefront.com/hackathon/beachshirts"
+)
+
+type ShoppingServer struct {
+	HostURL string
+	Router  *chi.Mux
+}
+
+func NewServer() *ShoppingServer {
+	r := chi.NewRouter()
+
+	server := &ShoppingServer{GlobalConfig.ShoppingHost, r}
+
+	r.Route("/shop", func(r chi.Router) {
+		r.Get("/menu", server.getShoppingMenu)
+		r.Post("/order", server.orderShirts)
+	})
+
+	return server
+}
+
+func (s *ShoppingServer) Start() error {
+	log.Println("Shopping Server listening @", s.HostURL)
+	return http.ListenAndServe(s.HostURL, s.Router)
+}
+
+func (s *ShoppingServer) getShoppingMenu(w http.ResponseWriter, r *http.Request) {
+	resp, err := callGetAllStyles()
+	if err != nil {
+		WriteError(w, "Failed to get menu!", http.StatusPreconditionFailed)
+	}
+	io.Copy(w, resp.Body)
+}
+
+func (s *ShoppingServer) orderShirts(w http.ResponseWriter, r *http.Request) {
+
+	RandSimDelay()
+
+	var order Order
+	json.NewDecoder(r.Body).Decode(&order)
+
+	if RAND.Float32() < GlobalConfig.SimFailShopping {
+		WriteError(w, "Failed to order shirts!", http.StatusServiceUnavailable)
+		return
+	}
+
+	resp, err := callMakeShirts(order)
+	if err != nil {
+		WriteError(w, err.Error(), http.StatusPreconditionFailed)
+		return
+	}
```

Comment:
Good practice to `defer resp.Body.Close()` if there is no error. Otherwise you might leak resources.